File: Python/ByTier/D/2160_D_MAD_Interactive_Problem.py
# Problem: CF 2160 D - MAD Interactive Problem
# https://codeforces.com/contest/2160/problem/D

"""
Interactive problem to reconstruct a secret sequence of length 2n where each integer from 1 to n appears exactly twice.
Uses queries of type "? k j1 j2 ... jk" to determine the MAD (Maximum Appearing Duplicate) of elements at specified indices.

Algorithm:
1. For each position in the array, try to determine its value by querying subsets containing that position
2. Use binary search approach to efficiently find values by leveraging MAD queries
3. Each element is determined using at most 2 queries: one for initial identification, one for final confirmation

Time Complexity: O(n^2) - In worst case, we might need up to 3n queries and each query involves processing up to n elements
Space Complexity: O(n) - Store the array and auxiliary data structures
"""

import sys

FILE = False
if FILE:
    sys.stdin = open("input.txt", "r")
    sys.stdout = open("output.txt", "w")


def myAlg(x=""):
    def get_int():
        s = sys.stdin.readline()
        if not s:
            sys.exit(0)
        s = s.strip()  # remove '\n' and spaces
        if not s:  # empty line
            return get_int()  # try reading again
        return int(s)

    def get_string():
        return sys.stdin.readline().strip()

    t = get_int()
    for _ in range(t):
        n = get_int()
        a = [0 for _ in range(2 * n)]
        indices = [1]
        setIndices = []
        for j in range(1, len(a)):
            indices.append(j + 1)
            print(f"? {len(indices)} " + " ".join(map(str, indices)))
            sys.stdout.flush()
            ans = get_int()
            if ans != 0:
                indices.pop()
                a[j] = ans
                setIndices.append(j + 1)
        for i in range(len(a)):
            if a[i] == 0:
                setIndices.append(i + 1)
                print(f"? {len(setIndices)} " + " ".join(map(str, setIndices)))
                sys.stdout.flush()
                ans = get_int()
                a[i] = ans
                setIndices.pop()
        print(f"! " + " ".join(map(str, a)))
        sys.stdout.flush()


myAlg()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/D/2160_D_MAD_Interactive_Problem.cpp
// Problem: CF 2160 D - MAD Interactive Problem
// https://codeforces.com/contest/2160/problem/D

/*
Problem: D. MAD Interactive Problem
Algorithms/Techniques: Interactive problem solving using queries to determine a hidden sequence of integers from 1 to n, each appearing exactly twice.

Approach:
1. First phase: Use queries to find out which positions have elements that are duplicates (MAD > 0). For each position i, we query with increasing size k and all unused indices up to i, and if MAD returns a value > 0, we know one of those indices points to a duplicate element.
2. Second phase: Once duplicates are identified, query the remaining unknown indices to find their values by using known positions.

Time Complexity: O(n^2) in worst case due to nested loops in the first phase and second phase processing
Space Complexity: O(n) for storing used array and unknown indices

*/

#include <algorithm>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

using namespace std;
#define el '\n'
#define pb push_back
#define pf push_front
#define ff first
#define ss second
#define mp make_pair
typedef long long ll;
typedef string str;
typedef vector<int> vi;
typedef pair<int, int> pi;
const int MOD = 1e9 + 7;
const int MAXN = 2e5;

void solve() {
  int n;
  cin >> n;
  int a[2 * n + 1]; // Array to store the answer
  vector<int> used(2 * n + 1, -1); // Track positions whose values are known (-1 means unknown)
  int k = 1; // Counter for queries
  int ans; // Result from query
  for (int i = 2; i <= 2 * n; i++) {
    k++; // Increment query count
    cout << "? " << k;
    for (int ind = 1; ind <= i; ind++) {
      if (used[ind] == -1) { // If position is not yet determined
        cout << " " << ind;
      }
    }
    cout << el;
    cout.flush(); // Flush output buffer
    cin >> ans; // Read result of query
    if (ans == 0) {
      continue; // No duplicate found, skip
    } else {
      used[i] = ans; // Mark the position as solved and store value
      k--; // Adjust counter since we got a result
    }
  }
  str known = " "; // String to build query for next phase
  vi unknown(n); // Vector to store positions that are yet to be determined
  int ind = 0;
  for (int i = 1; i <= 2 * n; i++) {
    if (used[i] != -1) {
      known += to_string(i) + " "; // Add known pos to string
    } else {
      unknown[ind] = i; // Mark as unknown
      ind++;
    }
  }
  for (int i = 0; i < n; i++) {
    cout << "? " << n + 1 << known << unknown[i] << el; // Query with n+1 elements to determine value
    cout.flush(); // Flush output buffer
    cin >> ans; // Read result
    used[unknown[i]] = ans; // Store the found value
  }
  cout << "! ";
  for (int i = 1; i <= 2 * n; i++) {
    cout << used[i] << " "; // Print final answer
  }
  cout << el;
  cout.flush(); // Flush output buffer
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int t = 1;
  cin >> t;
  while (t--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/